NGS2_WITNESS_Cycle1_difi_prep_exp2.R

```diff
@@ -0,0 +1,218 @@
+# DIFI
+####################################################################################################
+
+# Breadboard extract function by event type
+
+breadboard.extract <- function (eventType) {
+  breadboard.list <- lapply(exp2, function(x) {
+    roll <- subset(x, 
+                   event == eventType, 
+                   select = c('id', 
+                              'data.name', 
+                              'data.value')
+                   )
+    return(roll)
+    })
+  breadboard.list2 <- mapply(`[<-`, 
+                             breadboard.list, 
+                             'session', 
+                             value = names(breadboard.list), 
+                             SIMPLIFY = FALSE)
+  breadboard.events<-do.call(rbind,
+                             lapply(breadboard.list2,data.frame)
+                             )
+  breadboard.frame<-dcast(breadboard.events, 
+                          id + session ~ data.name, 
+                          value.var = 'data.value')
+  return(breadboard.frame)
+}
+
+# Extract DIFI scores
+
+DIFI.final<-breadboard.extract ("DIFI")
+
+# Compute individual probability of cooperation
+
+av.cooperation.exp2 <-aggregate(exp2_cooperation$decision0d1c,
+                                by=list(exp2_cooperation$playerid),
+                                FUN=mean,
+                                na.rm=TRUE)
+colnames(av.cooperation.exp2)<-c("pid", "prob.cooperate")
+
+### Create frame of nodes in network, with their final score and final group.
+
+# Final score
+
+score.exp2 <- breadboard.extract ("FinalScore")
+
+# Final group
+
+exp2_cooperation <- read.csv('NGS2-Cycle1-Experiment2/cooperation_exp2.csv',
+                             header = TRUE, sep = ',')
+
+final.group <- unique(subset (exp2_cooperation, 
+                       round_num == 15, 
+                       select = c('playerid', 'group')))
+
+colnames(final.group) <- c('pid', 'group')
+
+### Add network properties (cooperative structure)
+
+connections.list <- lapply(exp2, function(x) {
+  roll <- subset(x, 
+                 event == "Connected" | event == "Disconnected" , 
+                 select = c('id',
+                            "event",
+                            'data.name', 
+                            'data.value')
+  )
+  return(roll)
+})
+
+connections.list2 <- mapply(`[<-`, 
+                            connections.list, 
+                           'session', 
+                           value = names(breadboard.list), 
+                           SIMPLIFY = FALSE)
+connection.events<-do.call(rbind,
+                           lapply(connections.list2,data.frame)
+)
+connections.exp2.dcast <- dcast(connection.events,
+                                id+session+event ~ data.name,
+                                value.var="data.value")
+
+connections.exp2.final <- connections.exp2.dcast[0,]
+
+for (i in 1:length(connections.exp2.dcast$id))

```

Comment:
It would probably be a safe bet to make sure the data frame is truly ordered to ensure the connection/disconnections happening as you iterate over rows happen in the right order.

Reply:
OK, sorted by session and then id, that should put everything in chronological order: 

connections.exp2.dcast <- connections.exp2.dcast[order(connections.exp2.dcast$session, 
                                                       connections.exp2.dcast$id),]